src/plutus-core-execution.k

```diff
@@ -102,57 +90,85 @@ module PLUTUS-CORE-EXECUTION
   rule toString(B:ByStr):String
     => substrString(toStringVerbatim(B), 1, lengthString(toStringVerbatim(B)))
 
-  rule ( builtin take:Name ( N:Int B:ByStr ) )
+  rule ( builtin take:Name N:Int B:ByStr )
     => toByStr(substrString(toString(B), 0, N))
 
-  rule ( builtin drop:Name ( N:Int B:ByStr ) )
+  rule ( builtin drop:Name N:Int B:ByStr )
     => toByStr(substrString(toString(B), N, lengthString(toString(B))))
 
-  rule ( builtin concatenate:Name ( B1:ByStr B2:ByStr ) )
+  rule ( builtin concatenate:Name B1:ByStr B2:ByStr )
     => toByStr(toString(B1) +String toString(B2))
 
-  rule ( builtin sha3_256:Name ( B:ByStr ) ) => toByStr(Sha3256(toString(B)))
-  rule ( builtin sha2_256:Name ( B:ByStr ) ) => toByStr(Sha256(toString(B)))
+  rule ( builtin sha3_256:Name B:ByStr ) => toByStr(Sha3256(toString(B)))
+  rule ( builtin sha2_256:Name B:ByStr ) => toByStr(Sha256(toString(B)))
 
-  rule ( builtin equalsByteString:Name ( B1:ByStr B2:ByStr ) ) => #primTrue
+  rule ( builtin equalsByteString:Name B1:ByStr B2:ByStr ) => #primTrue
     when toString(B1) ==String toString(B2)
-  rule ( builtin equalsByteString:Name ( B1:ByStr B2:ByStr ) ) => #primFalse
+  rule ( builtin equalsByteString:Name B1:ByStr B2:ByStr ) => #primFalse
     when toString(B1) =/=String toString(B2)
 
   /***************************************************************************/
   /************************   LOCAL REDUCTIONS   *****************************/
   /***************************************************************************/
 
-  rule [ ( lam X M ) (N:Tm) ]  => M[N / X]
+  syntax K ::= #appFreeze(TmList)                                      [prefer]
+
+  // Explicit strictness for nested application
+  rule [ F:Val M1:Tm M2:Tm MS:TmList ] => [ F M1 ] ~> #appFreeze(M2 MS)
+  rule M:Val ~> #appFreeze(MS) => [ M MS ]
+  rule [ [ F:Val M:Tm ] MS:TmList ] => [ F M MS ]
+
+  // When a lambda abstracts more than one variable, desugar it to a
+  // combination of unary lambdas.
+  rule [ ( lam X1:Var X2:Var XS:VarList M:Tm ) N:Tm ]
+    => [ ( lam X1 .VarList ( lam X2 XS M ) ) N ]
+  // In the unary case, just perform the substitution.
+  rule [ ( lam X:Var .VarList M:Tm ) N:Tm ] => M[N / X]
+
```

Comment:
I still don't love how application is handled. I think we should try to handle it more like the nested lambdas in the sense that we desugar it to explicit unary application.